### ISSUE TITLE
Implement run styles

### ISSUE THREAD

**Issue by Gocnak** [momentum-mod/website#205]
Things like W-only, lowgrav, etc

**Comment by jason-e** [comment]
## Angle Restrictions

Names/terminology used here are placeholders and not necessarily official/final

### Group A (0 degree turn buffer)
- Normal (no key restrictions)
    - Maybe includes backwards as a subtag (need to figure out what the criteria for "backwards" is though)
- SW (W+S)
    - Could include left/right direction only as a subtag
- Surf HSW (WA+SD or WD+SA)
    - Could include backwards as a subtag
    
### Group B (90 degree turn buffer)
- Bhop HSW (WA+WD / SA+SD / WA+SA / WS+SD)
- Other (W+A / W+D / S+A / S+D) 
    - These aren't really a thing but it works the same and would go in this group
    
### Group C (180 degree turn buffer)
- Single key (W / A / S / D)    

## Notes
- Each group is its own leaderboard, and within each leaderboard, each run could be tagged with what specific restrictions that run claimed to have (the only thing that can be verified is which group the run should go in, beyond that the client can fake angles or whatever). You could filter to some specific tag(s) when browsing, but as far as ranking/points goes, there can be no finer breakdown than just these 3 groups.
- Each group is progressively more restrictive, so there could also be one collated leaderboard with all 3 groups and one with B+C as a way to see everything at a glance -- probably not for ranking/points themselves though; not sure how that would work vs the individual groups.
- Technically you could even have 45 deg (e.g. A+WD etc) and 135 deg (e.g. W+WA etc) groups but that would be a bit silly

### Regarding in-game behavior
- The player should be able to choose which of these to run as before they start and the game will apply the necessary restrictions
- Some community servers on other games don't impose angle/key restrictions initially, but determine which mode the player is trying to run based on their inputs. We could have this as a second method but it shouldn't be the only one.
- Basically the player should have the option of having their client explicitly impose restrictions when playing, but the final mode/tags determination would be done independently by the verification backend. Additional tags that the player didn't explicitly select could be added to the run at the time of verification if the run qualifies for them.

**Comment by jason-e** [comment]
## Complete mode hierarchy vs. what we should actually support

![image](https://user-images.githubusercontent.com/2745352/188754360-cc7cda62-4de8-4a6d-9145-48fcb5f2b521.png)

This flow chart visualizes which modes/categories nest into others. For example, since "Normal" runs are strictly harder than segmented runs, the segmented leaderboard _could_ include "normal" runs as well (either for ranking purposes or just casual comparison).

However we should probably not actually support ranking/viewing runs based on this entire hierarchy altogether:
- I have no idea how ranked points should work in such an ecosystem
- It would probably be exceedingly confusing to browse and gain meaningful insight from, even without considering ranked points
- Whether for ranked points or not, it could take attention away from niche modes -- for example a TAS leaderboard could include ALL runs in its hierarchy, but it would be very similar to the "normal" leaderboard where most runs will be concentrated.

We could start with just these core categories and keep them totally separate. They could be reorganized/grouped/separated later on if a better idea becomes apparent, especially once we see how things play out:
- Normal
- TAS
- Segmented
- Slow-Mo
- Low-Grav (probably only on some game modes)
- Anything else we come up with

Each of these leaderboards would be their own isolated sets that determine ranked points. Any run would belong to exactly one of these groups. They would have tags putting them in groups A/B/C as described above, but those subgroups would not have their own ranked points. Users could filter based on tags and see a "ranking" for just within those tags, but only for casual comparison. We could have an option to include all less-restrictive runs on the display (again just for a casual comparison), for example, such a toggle on the Segmented leaderboard would show all Normal runs as well.

**Comment by hexaflexahexagon** [comment]
FWIW the general consensus so far has been that _all_ styles give you zero rank just for simplicity. You get whatever relevant completion points (maybe 1.5x for A, 2x for B, 3x for C or something since they are more difficult, .75x for segmented and .5x for TAS or something) but the leaderboards give no ranked points at all and do not contribute to e.g. user global bhop rank. 

Also, IMO the organization of "assisted" / "challenge" run flags like TAS, segmented, low grav, etc should be completely separate from angles UX-wise. Maybe this does not make sense globally but in my mind (SJ/RJ perspective) they are really not similar enough to bunch into the same basket. It would make viewing leaderboards for example a mess unless I am lacking imagination for some sort of advanced filter UI.
